### PR TITLE
Fixed #32712 -- Deprecated django.utils.baseconv module.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32712

### CODE REVIEW

File: tests/utils_tests/test_baseconv.py [hunk 1]
```diff
@@ -1,50 +1,71 @@
...
+
         base7 = BaseConverter('cjdhel3', sign='g')
         self.assertEqual(repr(base7), '<BaseConverter: base7 (cjdhel3)>')
+
+
+class TestBaseConvDeprecationWarning(SimpleTestCase):
+    def test_deprecation_warning(self):
+        msg = 'The baseconv module is deprecated.'
+        with self.assertRaisesMessage(RemovedInDjango50Warning, msg):
+            from django.utils.baseconv import BaseConverter  # NOQA
```
Comment (hramezani) [inline tests/utils_tests/test_baseconv.py:71]: @felixxm Do you have any idea why this test fails with `AssertionError: RemovedInDjango50Warning not raised`?
It passes in my local setup and `django-docker-box`.

File: tests/utils_tests/test_baseconv.py [hunk 2]
```diff
@@ -1,50 +1,71 @@
...
+from django.utils.deprecation import RemovedInDjango50Warning
 
-from django.utils.baseconv import (
-    BaseConverter, base2, base16, base36, base56, base62, base64,
-)
 
-
-class TestBaseConv(TestCase):
+@ignore_warnings(category=RemovedInDjango50Warning)
+class TestBaseConv(SimpleTestCase):
```
Comment (felixxm) [inline tests/utils_tests/test_baseconv.py:6]: I would move all internal import to the module level, e.g.
```python
with ignore_warnings(category=RemovedInDjango50Warning):
    from django.utils.baseconv import (
        BaseConverter, base2, base16, base36, base56, base62, base64,
    )
 

class TestBaseConv(SimpleTestCase):
    ...
```

File: docs/releases/4.0.txt [hunk 3]
```diff
@@ -432,6 +432,8 @@ Miscellaneous
   :attr:`~django.test.TestCase.databases` with the
   :ref:`serialized_rollback <test-case-serialized-rollback>` option enabled.
 
+* The undocumented module ``django.contrib.baseconv`` is deprecated.
```
Comment (felixxm) [inline docs/releases/4.0.txt:435]:
```suggestion
* The undocumented ``django.utils.baseconv`` module is deprecated.
```

**Review by felixxm — COMMENTED** [review]
@hramezani Thanks :+1: Please add also a notes to the `docs/internals/deprecation.txt`.

File: django/core/signing.py [hunk 4]
```diff
@@ -59,6 +59,40 @@ class SignatureExpired(BadSignature):
...
+    while s > 0:
+        res = BASE62_ALPHABET[s % 62] + res
+        s //= 62
+
+    return sign + res
+
+
+def b62_decode(s):
+    s = str(s)
+    if str(s)[0] == '-':
```
Comment (smithdc1) [inline django/core/signing.py:81]: No need to convert s to a string again (and the same further down).

File: django/core/signing.py [hunk 5]
```diff
@@ -59,6 +59,40 @@ class SignatureExpired(BadSignature):
     pass
 
 
+def b62_encode(s):
+    sign = ''
+    if str(s)[0] == '-':
+        sign = '-'
+
+    s = abs(int(s))
```
Comment (smithdc1) [inline django/core/signing.py:67]: Could put this further up to return early for this case?

File: django/core/signing.py [hunk 6]
```diff
@@ -59,6 +59,40 @@ class SignatureExpired(BadSignature):
...
+        sign = -1
+    else:
+        sign = 1
+
+    if s == '0':
+        return 0
+
+    res = 0
+    for digit in str(s):
+        res = res * 62 + BASE62_ALPHABET.index(digit)
```
[the remark this answers is not in the document]
Comment (hramezani) [reply django/core/signing.py:92]: I don't know. AFAIK, it just used in [TimestampSigner](https://github.com/django/django/blob/65a9d0013d202447dd76a9cb3c939aa5c9d23da3/django/core/signing.py#L187).
IMHO, we can keep it like this.

File: django/core/signing.py [hunk 7]
```diff
@@ -59,6 +59,40 @@ class SignatureExpired(BadSignature):
...
 
+def b62_encode(s):
+    num = abs(int(s))
+    if num == 0:
+        return '0'
+
+    sign = ''
+    if str(s)[0] == '-':
+        sign = '-'
+
```
[the remark this answers is not in the document]
Comment (hramezani) [reply django/core/signing.py:70]: Yes, you are right. but the deprecated version in `baseconv` accepts `str` and `int` as input. That's why I convert it to str and then compare the first char.

File: django/core/signing.py [hunk 8]
```diff
@@ -59,6 +59,40 @@ class SignatureExpired(BadSignature):
...
+    res = ''
+    while num > 0:
+        res = BASE62_ALPHABET[num % 62] + res
+        num //= 62
+
+    return sign + res
+
+
+def b62_decode(s):
+    s = str(s)
```
[the remark this answers is not in the document]
Comment (hramezani) [reply django/core/signing.py:80]: The deprecated version accepts `int` and `str`. If we remove this line, this won't support `int`

File: django/core/signing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/core/signing.py:70]: Indeed, but this is a new, private copy intended for use in `django.core.signing` only. It seems sensible to limit it to the minimum required functionality.

**Comment by felixxm** [comment]
@hramezani Thanks for updates :+1:

**Pull request merged**